Лаб. 2 6411 Шехматов Виктор

Nakamo312 opened the pull request:
ФИО: Шехматов Виктор
Номер группы: 6411
Номер лабораторной: 2
Номер варианта: 25
Краткое описание предметной области: Поликлиника.
Краткое описание добавленных фич: На инфраструктурном уровне расположены репозитории, отвечающие за работу с данными. Они предоставляют абстракции для выполнения базовых операций с сущностями, таких как создание, чтение, обновление и удаление записей.  Прикладной уровень содержит сервисы, которые инкапсулируют основную бизнес-логику. Эти сервисы используют репозитории для взаимодействия с данными и добавляют дополнительную валидацию, трансформацию данных и сложные запросы. Также на этом уровне определены DTO-объекты для передачи данных между слоями и настроены профили AutoMapper для автоматического преобразования между доменными моделями и DTO.

Программно-интерфейсный слой представлен контроллерами, которые обрабатывают HTTP-запросы и взаимодействуют с сервисами прикладного уровня. Контроллеры отвечают за маршрутизацию, сериализацию данных и возврат соответствующих HTTP-ответов. Дополнительно в этом слое реализовано промежуточное программное обеспечение для логирования запросов, которое может быть расширено для решения других задач, таких как аутентификация, обработка ошибок или кэширование.

Nakamo312 committed:
initial commit

Nakamo312 committed:
fix test.yml

Nakamo312 committed:
add comments for IModel.cs

Nakamo312 committed:
change field IsRepeated in Appointment data model

Nakamo312 committed:
added Repository class. Added DataBase context class

Nakamo312 committed:
create Application Layer: -Added MappingProfile;\n-Added DTOs;

Nakamo312 committed:
created Base Controller and Services

Nakamo312 committed:
modified DTOs

Nakamo312 committed:
Add API Controllers

Nakamo312 committed:
Add InMmempry Repositories

Nakamo312 committed:
Added API CRUD endpoints, added DTO representation schemes, addrd logging through middleware

Nakamo312 committed:
update README.md

Nakamo312 committed:
changes status code and response header in Create methods

Nakamo312 committed:
Added analytic services and controllers

Nakamo312 committed:
rm file

Nakamo312 committed:
fix test.yml

Nakamo312 committed:
fix test.yml

Nakamo312 committed:
remove exceptions messages from endpoints

Nakamo312 committed:
change swager gen documentation for DTO schemes. Change content-type for all endpoints

danlla commented on Clinic/Clinic.Api/Clinic.Api.csproj:
Название проекта (и соответственно папки в которой он лежит) лучше сделать `Clinic.Api`

danlla commented on Clinic/Clinic.API/Clinic.API.csproj:
```diff
@@ -0,0 +1,27 @@
+    <PackageReference Include="Aspire.MongoDB.Driver.v3" Version="9.5.1" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="9.0.10" />
+    <PackageReference Include="Swashbuckle.AspNetCore" Version="6.6.2" />
+    <PackageReference Include="SyrianBallaS.AutoMapper.Extensions.Microsoft.DependencyInjection.Signed" Version="3.2.0" />
```
`Aspire.MongoDB.Driver.v3`
`Microsoft.EntityFrameworkCore.InMemory`
`SyrianBallaS.AutoMapper.Extensions.Microsoft.DependencyInjection.Signed`

не нужны и никак не используются в этом проекте

danlla commented on Clinic/Clinic.API/Clinic.API.csproj:
```diff
@@ -0,0 +1,27 @@
+  <ItemGroup>
+    <ProjectReference Include="..\Clinic.Application\Clinic.Application.csproj" />
+    <ProjectReference Include="..\Clinic.Domain\Clinic.Domain.csproj" />
+    <ProjectReference Include="..\Clinic.Infrastructure\Clinic.Infrastructure.csproj" />
```
`Clinic.Application` уже содержит референс на `Clinic.Domain` и `Clinic.Infrastructure`

danlla commented on Clinic/Clinic.Api/Clinic.Api.csproj:
```diff
@@ -0,0 +1,27 @@
+
+  <ItemGroup>
+    <Folder Include="Middleware\" />
+  </ItemGroup>
```
Это можно убрать

danlla commented on Clinic/Clinic.API/Controllers/AnalyticController.cs:
```diff
@@ -0,0 +1,146 @@
+using Clinic.Application.Services;
+using Microsoft.AspNetCore.Mvc;
+
+namespace analytic.API.Controllers;
```
Неправильный namespace

danlla commented on Clinic/Clinic.API.Host/Controllers/AnalyticController.cs:
```diff
@@ -0,0 +1,146 @@
+[Route("api/[controller]")]
+public class AnalyticController(IAnalyticQueryService analyticQueryService) : ControllerBase
+{
+    private readonly IAnalyticQueryService _analyticQueryService = analyticQueryService;
```
Можно везде сразу использовать `analyticQueryService`, данное поле здесь не нужно

Проверить в остальных классах

danlla commented on Clinic/Clinic.API.Host/Controllers/AnalyticController.cs:
```diff
@@ -0,0 +1,146 @@
+        }
+        catch (InvalidOperationException)
+        {
+            return StatusCode(500);
```
Не совсем user-friendly возвращать просто 500 код без описания ошибки, ну да ладно, можно и так оставить

Nakamo312 replied:
А здесь я долго колебался, стоит ли возвращать пользователю какую то информацию, и подумал  что из соображения безопасности лучше не уточнять причину ошибки

danlla replied:
Саму причину можно уточнить, тем более если она связана с тем, что пользователь передал какие-то неправильные данные

А так да, лишнюю информацию, в том числе и полное описание исключения передавать не стоит

danlla commented on Clinic/Clinic.API.Host/Controllers/CrudControllerBase.cs:
```diff
@@ -0,0 +1,158 @@
+{
+    protected readonly ICrudService<TDto, TCreateDto, TUpdateDto> _service = service;
+
+    protected virtual string EntityName => GetType().Name.Replace("Controller", "");
```
Вроде как нигде не используется

danlla commented on Clinic/Clinic.Api/Middleware/LoggingMiddleware.cs:
Излишний класс для лабораторной, но если хочется чтобы формат логов был таким, то ладно

Nakamo312 replied:
Я его добавил скорее исключительно чтоб потом можно было менять логику логирования ошибок в случае исключений для которых 500. Но на данный момент да, от него не сильно много смысла

danlla commented on Clinic/Clinic.API/Program.cs:
```diff
@@ -0,0 +1,134 @@
+    LoggerFactory.Create(builder => builder.AddConsole()));
+
+IMapper? mapper = mapperConfig.CreateMapper();
+builder.Services.AddSingleton(mapper);
```
Можно вот так:

```csharp
builder.Services.AddAutoMapper(config =>
{
    config.AddProfile(new MappingProfile());
    config.AddProfile(new AnalyticQueryProfile());
});
```

Не уверен что тут необходимо прокидывает фабрику для логгеров

danlla commented on Clinic/Clinic.API/Program.cs:
```diff
@@ -0,0 +1,134 @@
+    provider => new BaseCrudService<Patient, PatientResponseDto, PatientCreateDto, PatientUpdateDto>(
+        provider.GetRequiredService<IRepository<Patient>>(),
+        provider.GetRequiredService<IMapper>()
+    ));
```
Здесь можно проще, для этого DI и существует:

```csharp
builder.Services.AddScoped<ICrudService<PatientResponseDto, PatientCreateDto, PatientUpdateDto>, BaseCrudService<Patient, PatientResponseDto, PatientCreateDto, PatientUpdateDto>>();
```

ниже аналогично

danlla commented on Clinic/Clinic.API/Program.cs:
```diff
@@ -0,0 +1,134 @@
+    {
+        await appointmentRepo.CreateAsync(appointment);
+    }
+}
```
Лучше инициализацию обернуть в отдельный сервис, который через DI получит нужные ему сервисы
И здесь получать этот сервис и вызывать его один метод

danlla commented on Clinic/Clinic.Application/Dtos/Doctors/DoctorCreateDto.cs:
```diff
@@ -0,0 +1,42 @@
+    /// <summary>
+    /// Year of birth of the doctor.
+    /// </summary>
+    [Required(ErrorMessage = "Year of birth is required")]
+    [Range(1900, 2100, ErrorMessage = "Year of birth must be between 1900 and 2100")]
+    public required uint YearOfBirth { get; set; }
```
Предложил бы здесь использовать еще атрибут `[DefaultValue()]`, чтобы задать дефолтное значение, которое будет отображаться в сваггере, а то сейчас он показывает 2100 при создании, было бы удобнее если бы это было значение больше похожее на правду

Можно и в остальных проперти это сделать, но это не замечание

danlla commented on Clinic/Clinic.Application/DTOs/Doctors/DoctorResponseDto.cs:
```diff
@@ -0,0 +1,34 @@
+/// <summary>
+/// Data transfer object representing a patient in response.
+/// </summary>
+public class DoctorResponseDto : Model
```
Тут можно было сразу от `Doctor` отнаследоваться и не писать свойства

Ну и во всех дто на получение так же

danlla commented on Clinic/Clinic.Application/DTOs/Analytics/DoctorsExperienceQueryDto.cs:
```diff
@@ -0,0 +1,12 @@
+    /// Minimum years of experience
+    /// </summary>
+    public uint MinExperience { get; set; }
+}
```
Не уверен, что есть смысл в этом классе, в метод контроллера можно просто прокидывать `uint` параметр из запроса

danlla commented on Clinic/Clinic.Application/DTOs/Analytics/IdsResponseDto.cs:
```diff
@@ -0,0 +1,12 @@
+    /// Collection of IDs
+    /// </summary>
+    public IEnumerable<uint> Ids { get; set; } = new List<uint>();
+}
```
С этим классом думаю так же

danlla commented on Clinic/Clinic.Application/DTOs/Analytics/NamesResponseDto.cs:
```diff
@@ -0,0 +1,12 @@
+    /// Collection of names
+    /// </summary>
+    public IEnumerable<string> Names { get; set; } = new List<string>();
+}
```
И с этим, ну и со всеми похожими

danlla commented on Clinic/Clinic.Application/Dtos/Patients/PatientCreateDto.cs:
```diff
@@ -0,0 +1,66 @@
+    /// <summary>
+    /// Gender of the patient.
+    /// </summary>
+    [Required(ErrorMessage = "Gender is required")]
+    [EnumRange(typeof(Gender))]
+    public required Gender Gender { get; set; }
```
Из не очень удобного: пользователю апи нужно при создании задавать для enum числовые значения, который он не знает как мапятся на реальные значение

В принципе можно это не исправлять

danlla commented on Clinic/Clinic.Application/DTOs/Patients/PatientCreateDto.cs:
```diff
@@ -0,0 +1,66 @@
+    /// </summary>
+    [Phone(ErrorMessage = "Invalid phone number format")]
+    [StringLength(20, ErrorMessage = "Phone number must not exceed 20 characters")]
+    public string? ContactPhone { get; set; }
```
Если есть определенный формат, который сервис ожидает, то тут лучше тогда навесить атрибут `RegularExpression`

danlla commented on Clinic/Clinic.Application/DTOs/Patients/PatientResponseDto.cs:
```diff
@@ -0,0 +1,50 @@
+    /// <summary>
+    /// Gender of the patient.
+    /// </summary>
+    public required Gender Gender { get; set; }
```
Здесь кстати тоже не очень удобно, что пользователь апи получается вместо понятных названий число из enum, но это не так важно

danlla commented on Clinic/Clinic.Application/DTOs/Analytics/RepeatedAppointmentsQueryDto.cs:
```diff
@@ -0,0 +1,18 @@
+    /// End date of the range
+    /// </summary>
+    public DateTime EndDate { get; set; }
+}
```
Этот клас так же выглядит как ненужный

Если все подобные классы нужны для комментариев, то сваггер возьмет комментарий из `summary` с описанием параметров метода

danlla commented on Clinic/Clinic.Application/DTOs/Analytics/RoomAppointmentsQueryDto.cs:
```diff
@@ -0,0 +1,22 @@
+    /// Month end date
+    /// </summary>
+    public DateTime MonthEnd { get; set; }
+}
```
И этот класс аналогично

danlla commented on Clinic/Clinic.Application/Profiles/AnalyticQueryProfile.cs:
```diff
@@ -0,0 +1,25 @@
+        CreateMap<DoctorPatientsQueryDto, DoctorPatientsQueryDto>();
+        CreateMap<RepeatedAppointmentsQueryDto, RepeatedAppointmentsQueryDto>();
+        CreateMap<RoomAppointmentsQueryDto, RoomAppointmentsQueryDto>();
+    }
```
В маппинге класса на него же нет никакого смысла

danlla commented on Clinic/Clinic.Application/Profiles/MappingProfile.cs:
```diff
@@ -0,0 +1,78 @@
+
+        CreateMap<SpecializationUpdateDto, Specialization>()
+            .ForAllMembers(opts => opts.Condition((src, dest, srcMember) =>
+                !IsDefaultOrEmptyValue(srcMember)));
```
Если это делалось для того, чтобы не перезаписывать id связанных сущностей при изменений, то идея неплохая, но есть проблема в таком решении, сейчас нельзя задать значение `0` например для опыта врача

В принципе автомаппер позволяет прописывать правила для каждого свойства и можно явно указать как маппить id связанных сущностей если они например `null` в дто на обновления

danlla commented on Clinic/Clinic.Application/Services/AnalyticQueryService.cs:
```diff
@@ -0,0 +1,136 @@
+    private readonly IMapper _mapper = mapper;
+
+    /// <inheritdoc/>
+    public async Task<NamesResponseDto> GetDoctorsWithExperienceAsync(DoctorsExperienceQueryDto queryDto)
```
Здесь возвращаются только имена, здесь нужно возвращать полностью врачей, то есть `DoctorResponseDto`

В методах ниже так же

danlla commented on Clinic/Clinic.Application/Services/BaseCrudService.cs:
```diff
@@ -0,0 +1,131 @@
+    where TUpdateDto : class
+{
+    protected readonly IRepository<TModel> _repository = repository;
+    protected readonly IMapper _mapper = mapper;
```
Эти поля не нужны

danlla commented on Clinic/Clinic.Application/Services/IAnalyticQueryService.cs:
```diff
@@ -0,0 +1,47 @@
+    /// </summary>
+    /// <param name="queryDto">Query parameters</param>
+    /// <returns>List of doctor full names</returns>
+    Task<NamesResponseDto> GetDoctorsWithExperienceAsync(DoctorsExperienceQueryDto queryDto);
```
Нужно во всех определениях методов во всех интерфейсах проставить `public`

danlla commented on Clinic/Clinic.Infrastructure/Properties/AssemblyInfo.cs:
Предполагаю, что этот файл тут лишний и его можно убрать

danlla commented on Clinic/Clinic.Application/Validation/EnumRangeAttribute.cs:
```diff
@@ -0,0 +1,39 @@
+    public EnumRangeAttribute(Type enumType)
+    {
+        _enumType = enumType;
+    }
```
primary constructor

danlla commented on Clinic/Clinic.Domain/Models/Abstract/Model.cs:
```diff
@@ -0,0 +1,12 @@
+﻿namespace Clinic.Domain.Models;
```
Неправильный namespace

А еще этот класс лежит в папке `Interfaces` хотя интерфейсом не является, возможно стоит переименовать папку

danlla replied:
До сих пор неправильный namespace

danlla commented on Clinic/Clinic.Infrastructure/Repositories/InMemoryRepository.cs:
Опечатка в названии файла

danlla commented on Clinic/Clinic.Infrastructure/Repositories/InMmemoryRepository.cs:
```diff
@@ -0,0 +1,78 @@
+/// <typeparam name="T">The type of entity this repository works with, must inherit from Model.</typeparam>
+public class InMemoryRepository<T> : IRepository<T> where T : Model
+{
+    private readonly List<T> _entities = new();
```
```csharp
private readonly List<T> _entities = [];
```

danlla commented on Clinic/Clinic.Infrastructure/Repositories/InMmemoryRepository.cs:
```diff
@@ -0,0 +1,78 @@
+        if (existing != null)
+        {
+            _entities.Remove(existing);
+            _entities.Add(entity);
```
При обновлении удалять, а потом заново создавать не очень хорошая идея, запрос будет работать дольше и съедать больше ресурсов, лучше обновлять сущность

danlla commented on Clinic/Clinic.Tests/LinqQueryTests.cs:
```diff
@@ -0,0 +1,145 @@
+    [Fact]
+    public void GetPatientsInfo()
+    {
+        var currentDate = DateTime.Now;
```
Лучше убрать `DateTime.Now` из тестов

danlla reviewed (changes requested):
1. Дто стало много, лучше разбить их на подпапки по сущностям
2.  Папку `DTOs` лучше переименовать в `Dtos` и потом поправить namespace

Nakamo312 committed:
refactor: decouple API layer from domain models

- Remove Domain.Model dependency from base controller
- Use reflection for ID access in DTOs instead of Model inheritance
- Keep API layer dependent only on DTOs and application contracts

Nakamo312 committed:
Remarks have been fixed

Nakamo312 committed:
Remarks have been fixed

Nakamo312 committed:
Remarks have been fixed

Nakamo312 commented:
Я столкнулся с дилеммой: с одной стороны, использование generic-компонентов (сервисов и контроллеров) позволяло избежать дублирования кода, но при этом в API слое появлялись зависимости от доменных моделей, что мне сильно не нравилось. С другой стороны, создание конкретных классов-наследников для каждого типа сущности устранило зависимости от доменных моделей в API слое, но привело к написанию большого количества почти пустых классов, что свело на нет преимущества generic-подхода. Как лучше всего поступить?

Nakamo312 committed:
Remarks have been fixed

Nakamo312 committed:
fix regex in PatientUpdateDto

Nakamo312 committed:
Improve error management: hide detailed validation errors

Nakamo312 committed:
revert last changes

danlla commented:
> Я столкнулся с дилеммой: с одной стороны, использование generic-компонентов (сервисов и контроллеров) позволяло избежать дублирования кода, но при этом в API слое появлялись зависимости от доменных моделей, что мне сильно не нравилось. С другой стороны, создание конкретных классов-наследников для каждого типа сущности устранило зависимости от доменных моделей в API слое, но привело к написанию большого количества почти пустых классов, что свело на нет преимущества generic-подхода. Как лучше всего поступить?

В этом коде достаточно много абстракций, которые мне кажутся излишними для данной работы, но стремление сделать все максимально абстрактно и на интерфейсах похвально, предлагаю оставить как есть

Большое количество почти пустых классов это не так плохо, если речь про дто на чтение, то возможна ситуация, когда в доменном классе появятся новые проперти, которые не должны будут возвращаться из апи и нужно будет просто описать дто на чтение с теми данными, которые он может возвращать, а остальную логику не нужно будет трогать


Я смотрю вы после запроса на ревью делаете еще изменения, лучше так не делать, сделайте все изменения и только потом отправляйте запрос на ревью, отпишите комментом тогда здесь, когда закончите изменения

Nakamo312 commented:
На данный момент по изменениям это все

Nakamo312 commented:
А нет, еще вопрос: как лучше оставить под аналитику отдельный контроллер - или вынести логику в контроллеры для дефолтных сущностей добавив в методы GetAll дополнительные необязательные query params? Во втором варианте это более CRUD ориентировано, и еще появляется возомжность композиции аналитических фильтров фильтр

danlla commented:
> А нет, еще вопрос: как лучше оставить под аналитику отдельный контроллер - или вынести логику в контроллеры для дефолтных сущностей добавив в методы GetAll дополнительные необязательные query params?

Можете оставить как есть

Nakamo312 commented:
Еще пытался изменить ситуацию с тем - что при возникновении ошибок дессериализации JSON, сообщения возвращаемые пользователю раскрывают структуру слоев приложения, я пробовал отключить в Program.cs это поведение - заменив кастомныv обработчиком - но тогда у меня и удобные сообщения об ошибках валидации заменялись слишком общими описаниями из кастомного обработчика, и мне показалось что то как я делаю это миддлварями - это колхоз и можно как то проще. (логировать каждый метод вручную очень нехочется, хочется сделать это более централизовано)

Nakamo312 committed:
change response stats code in DELETE methods

Nakamo312 commented:
изменений больше не будет до следующего ревью

danlla commented on Clinic/Clinic.Api/Clinic.API.Host.http:
Этот файл можно удалить

danlla commented on Clinic/Clinic.Api/Clinic.Api.csproj:
```diff
@@ -0,0 +1,24 @@
+
+  <ItemGroup>
+    <Folder Include="Middleware\" />
+  </ItemGroup>
```
Это все еще можно убрать

Nakamo312 replied:
Не могу понять почему эта вещь сама добавляется при пересборке проекта. А если ее убирать - то он требует использования using Clinic.Api.Middleware в Program.cs

danlla replied:
Использование `using Clinic.Api.Middleware` это нормальное поведение

danlla commented on Clinic/Clinic.Api/Controllers/AnalyticController.cs:
```diff
@@ -0,0 +1,142 @@
+using Clinic.Application.Services;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Clinic.API.Host.Controllers;
```
namespace поменялся, это хорошо, но он до сих пор не соответствует действительности, это плохо

если вы работаете в visual studio, то она автоматом предложит поправить, а также выдаст ошибку 
<details>
<summary>Пруф</summary>
<img width="872" height="133" alt="image" src="https://github.com/user-attachments/assets/36e8ae97-75dd-4a4a-b862-2fc682f69266" />
</details>

если не в visual studio, то namespace должен содержать в себе название проекта и все папки в том порядке, в котором их нужно открыть чтобы добраться до файла, для которого пишется namespace

то есть здесь он будет: `Clinic.Api.Controllers`

danlla commented on Clinic/Clinic.Api/Controllers/AnalyticController.cs:
```diff
@@ -0,0 +1,142 @@
+    [ProducesResponseType(typeof(IEnumerable<uint>), 200)]
+    [ProducesResponseType(400)]
+    [ProducesResponseType(500)]
+    public async Task<ActionResult<IEnumerable<uint>>> GetRepeatedAppointments([FromQuery] DateTime startDate, [FromQuery] DateTime endDate)
```
Здесь до сих пор возвращается список id, а не полноценные сущности

Также хочу заметить, что в задании здесь нужно вычислять **_за последний месяц_**, то есть `startDate` и `endDate` не нужны

Можно их оставить, но сделать необязательными и если они не заданы, то вычислять их так, чтобы они обозначали последний месяц

danlla commented on Clinic/Clinic.Api/Controllers/AnalyticController.cs:
```diff
@@ -0,0 +1,142 @@
+    [ProducesResponseType(typeof(IEnumerable<uint>), 200)]
+    [ProducesResponseType(400)]
+    [ProducesResponseType(500)]
+    public async Task<ActionResult<IEnumerable<uint>>> GetAppointmentsForRoom([FromQuery] RoomAppointmentsQueryDto queryDto)
```
А для этого запроса в задании сказано **_за текущий месяц_**

danlla commented on Clinic/Clinic.Application/Clinic.Application.csproj:
```diff
@@ -0,0 +1,21 @@
+
+  <ItemGroup>
+    <ProjectReference Include="..\Clinic.Domain\Clinic.Domain.csproj" />
+    <ProjectReference Include="..\Clinic.Infrastructure\Clinic.Infrastructure.csproj" />
```
`Clinic.Infrastructure` уже ссылается на `Clinic.Domain`, значит можно отсюда убрать `Clinic.Domain`

danlla commented on Clinic/Clinic.Application/Services/AnalyticQueryService.cs:
```diff
@@ -0,0 +1,129 @@
+/// <summary>
+/// Implementation of clinic query service.
+/// </summary>
+public class AnalyticQueryService
```
Нет namespace

danlla commented on Clinic/Clinic.Infrastructure/Clinic.Infrastructure.csproj:
```diff
@@ -0,0 +1,23 @@
+
+  <ItemGroup>
+    <Folder Include="Properties\" />
+  </ItemGroup>
```
Это можно убрать

danlla commented on Clinic/Clinic.Infrastructure/Persistence/AppDbContext.cs:
Название файла отличается от названия класса

danlla commented on Clinic/Clinic.Infrastructure/Repositories/Interfaces/IRepository.cs:
```diff
@@ -0,0 +1,43 @@
+﻿namespace Clinic.Infrastructure.Repositories;
```
Неправильный namespace

danlla commented on Clinic/Clinic.Tests/Clinic - Backup.Tests.csproj:
Подобные файлы не должны попадать в репозиторий, его стоит убрать

danlla commented on Clinic/Clinic.Tests/Clinic.Tests.csproj:
```diff
@@ -0,0 +1,34 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\Clinic.Application\Clinic.Application.csproj" />
```
Тесты не должны зависеть от `Clinic.Application`

Для этого данные для тестов должны быть в `Clinic.Domain`, а в `LinqQueryTests` использовать `IClassFixture<DataSeed>`

danlla commented on Clinic/Clinic.Tests/DataFixture.cs:
То есть от этого файла можно в принципе избавиться, тем более к нему есть много замечаний, как минимум от студии:
<img width="1663" height="325" alt="image" src="https://github.com/user-attachments/assets/b615aef0-34da-4769-8a22-bebdfd77378a" />

danlla commented on Clinic/Clinic.Tests/LinqQueryTests.cs:
```diff
@@ -0,0 +1,145 @@
+/// </summary>
+public class LinqQueryTests(DataFixture testData) : IClassFixture<DataFixture>
+{
+    private readonly DataFixture _testData = testData;
```
Это поле не нужно

danlla commented on Clinic/Clinic.Infrastructure/Data/InMemoryDataSeeder.cs:
```diff
@@ -0,0 +1,76 @@
+        foreach (var patient in DataSeed.Patients)
+        {
+            await patientRepository.CreateAsync(patient);
+        }
```
Можно записать покороче:

```csharp
DataSeed.Patients.ForEach(p => patientRepository.CreateAsync(p))
```

Ну и ниже соответственно

Nakamo312 replied:
ForEach у List<T> не поддерживает await внутри лямбды

danlla replied:
Поддерживает, если правильно объявить лямбду:

```csharp
data.Patients.ForEach(async (p) => await patientRepository.CreateAsync(p));
```

Так как сделано сейчас, через `Select`, выглядит немного костыльно

Nakamo312 replied:
я имел в виду то, что ```List<T>.ForEach```  не дожидается асинхроных делегатов, потому что он возвращает ```void```, по итогу весь метод становится синхронным и я получаю Warning
<img width="1180" height="421" alt="изображение" src="https://github.com/user-attachments/assets/d77b0102-a353-411f-8109-1ce037424d5e" />

надеюсь такой вариант устроит, хотя это тоже самое что LINQ-метод ```Select``` только топорнее
```C#
var tasks = new List<Task>();

data.Patients.ForEach(p => tasks.Add(patientRepository.CreateAsync(p)));
data.Specializations.ForEach(s => tasks.Add(specializationRepository.CreateAsync(s)));
data.Doctors.ForEach(d => tasks.Add(doctorRepository.CreateAsync(d)));
data.Appointments.ForEach(a => tasks.Add(appointmentRepository.CreateAsync(a)));

await Task.WhenAll(tasks);
```

danlla replied:
Согласен, да можно оставить такой вариант, в принципе варианты с `Select` и с обычным `foreach` тогда также выглядят неплохо

danlla reviewed (changes requested):
Некоторые замечания не исправлены с прошлого ревью

Папка DTOs не переименовалась, возможно вы пытались ее переименовать и не вышло, потому что названия папок регистронезависимые в гите, чтобы поменять название нужно поменять его сначала на что-то совершенно новое (с учетом регистронезависимости), а потом поменять на нужное название

danlla commented on Clinic/Clinic.Application/DTOs/Patients/PatientCreateDto.cs:
```diff
@@ -0,0 +1,68 @@
+    /// <summary>
+    /// Contact phone number of the patient.
+    /// </summary>
+    [RegularExpression(@"^((8|\+7)[\- ]?)?(\(?\d{3}\)?[\- ]?)?[\d\- ]{7,10}$", ErrorMessage = "Phone number must be in the format '+7 (XXX) XXX-XX-XX'")]
```
вот такой номер: `8(8693 3-7 8614` будет считаться валидным для такой регулярки

вот такая регулярка вроде получше будет:

`^(\+?7|8)?[\s\-]?\(?(\d{3})\)?[\s\-]?(\d{3})[\s\-]?(\d{2})[\s\-]?(\d{2})$`

Nakamo312 replied:
`+7 908)6812697` вот такая сматчилась под эту регулярку

danlla replied:
Тогда вот такая:
```regex
^(\+?7|8)?[\s\-]?(?:\((\d{3})\)|(\d{3}))[\s\-]?(\d{3})[\s\-]?(\d{2})[\s\-]?(\d{2})$
```

Nakamo312 replied:
оставлю такую
```^(\+7|8)[\s]?(?:(?:\(\d{3}\)|\d{3}))(?:(?:[\s\-]?\d{3})(?:[\s\-]?\d{4}|(?:[\s\-]?\d{2}){2}))$```
очень неплохо вроде работает. Да и я думаю нужно будет в любом случае хранить в БД в одном формате все номера телефонов без лишних символов  - так просто экономнее. Как будто эта регулярка просто чтоб упростить парсинг дальнейший.

Nakamo312 committed:
comments corrected

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
rename folder

Nakamo312 committed:
registred DataSeed in DI container

Nakamo312 committed:
update phone number regex

Nakamo312 committed:
added phone number normalizatiom; change contracts for PUT methods

danlla commented on Clinic/Clinic.Api/Controllers/AnalyticController.cs:
```diff
@@ -0,0 +1,148 @@
+using Clinic.Application.Services;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Clinic.API.Controllers;
```
namespace до сих пор неверный

danlla commented on Clinic/Clinic.Domain/Models/Abstract/Model.cs:
```diff
@@ -0,0 +1,12 @@
+﻿namespace Clinic.Domain.Abstract.Models;
```
И здесь тоже

danlla reviewed (approved):
Поправить при сдаче следующей лабы
